packages/mui-material/src/useTouchRipple/useLazyRipple.ts

```diff
@@ -0,0 +1,107 @@
+'use client';
+import * as React from 'react';
+import useLazyRef from '@mui/utils/useLazyRef';
+import { TouchRippleActions } from '../ButtonBase/TouchRipple';
+
+type ControlledPromise<T = unknown> = Promise<T> & {
+  resolve: Function;
+  reject: Function;
+};
+
+/**
+ * Lazy initialization container for the Ripple instance. This improves
+ * performance by delaying mounting the ripple until it's needed.
+ */
+export class LazyRipple {
+  /** React ref to the ripple instance */
+  ref: React.MutableRefObject<TouchRippleActions | null>;
+
+  /** If the ripple component should be mounted */
+  shouldMount: boolean;
+
+  /** Promise that resolves when the ripple component is mounted */
+  private mounted: ControlledPromise | null;
+
+  /** If the ripple component has been mounted */
+  private didMount: boolean;
+
+  /** React state hook setter */
+  private setShouldMount: React.Dispatch<boolean> | null;
+
+  static create() {
+    return new LazyRipple();
+  }
+
+  constructor() {
+    this.ref = { current: null };
+    this.mounted = null;
+    this.didMount = false;
+    this.shouldMount = false;
+    this.setShouldMount = null;
+  }
+
+  mount() {
+    if (!this.mounted) {
+      this.mounted = createControlledPromise();
+      this.shouldMount = true;
+      this.setShouldMount!(this.shouldMount);
+    }
+    return this.mounted;
+  }
+
+  mountEffect = () => {
+    if (this.shouldMount && !this.didMount) {
+      Promise.resolve().then(() => {
+        if (this.ref.current !== null) {
+          this.didMount = true;
+          this.mounted!.resolve();
+        }
+      });
+    }
+  };
+
+  render() {
```

Comment:
It felt cleaner to me to encapsulate all the logic inside the class.

Reply:
I worry that by not following the [rules of hooks](https://react.dev/warnings/invalid-hook-call-warning#breaking-rules-of-hooks), we might introduce weird behavior, either now or in future React releases.

Reply:
We are following the rules of hooks, which rule are you concerned about? Maybe it would help if I rename the function to `use()`? That would highlight that it's really just a hook, just organized differently. The eslint react plugin is not implemented properly, it's triggering a false positive here as if it was detecting a class component. This isn't a class component.

Reply:
> which rule are you concerned about?

About calling hooks at the top level in the body of a custom hook and [not a nested function](https://react.dev/reference/rules/rules-of-hooks#:~:text=inside%20loops%2C%20conditions%2C-,nested%20functions,-%2C%20or%20try/), which is what I understand we're doing here.

> Maybe it would help if I rename the function to use()?

I'm not sure. The React documentation mentions that hooks shouldn't be called in regular Javascript functions: https://react.dev/reference/rules/rules-of-hooks#only-call-hooks-from-react-functions. I don't know if that includes a class method or not.

I think by moving this into `useLazyRipple`'s body, we're sticking to what React expects and avoiding any weird behavior.

Reply:
The `render()` function *is* a custom hook, not a nested function. Consider if we rename it to `use` to match hook naming rules, these two examples are completely equivalent:

```js
class LazyRipple {
  use() {
    const [a, b] = useState(c)
  }
}

// vs

function use(lazy: LazyRipple) {
  const [a, b] = useState(c)
}
```

But having the function live inside the class prevents breaking encapsulation. I have moved stuff around to make it more apparent that the logic is just a hook, could you take a loot at the change?

Reply:
Thanks for the explanation ✅